src/ripple/app/ledger/impl/TimeoutCounter.h

```diff
@@ -0,0 +1,101 @@
+//------------------------------------------------------------------------------
+/*
+    This file is part of rippled: https://github.com/ripple/rippled
+    Copyright (c) 2012, 2013 Ripple Labs Inc.
+
+    Permission to use, copy, modify, and/or distribute this software for any
+    purpose  with  or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#ifndef RIPPLE_APP_LEDGER_TIMEOUTCOUNTER_H_INCLUDED
+#define RIPPLE_APP_LEDGER_TIMEOUTCOUNTER_H_INCLUDED
+
+#include <ripple/app/main/Application.h>
+#include <ripple/beast/clock/abstract_clock.h>
+#include <ripple/beast/utility/Journal.h>
+#include <boost/asio/basic_waitable_timer.hpp>
+#include <mutex>
+
+namespace ripple {
+
+/**
+    This class is an "active" object. It maintains its own timer
+    and dispatches work to a job queue. Implementations derive
+    from this class and override the abstract hook functions in
+    the base.
+
+    This class is split from the PeerSet class.  // TODO remove after PR
+*/
+class TimeoutCounter
+{
+protected:
+    using ScopedLockType = std::unique_lock<std::recursive_mutex>;
+
+    TimeoutCounter(
+        Application& app,
+        uint256 const& hash,
+        std::chrono::milliseconds interval,
+        beast::Journal journal);
+
+    virtual ~TimeoutCounter() = 0;
+
+    /** Hook called from invokeOnTimer(). */
+    virtual void
+    onTimer(bool progress, ScopedLockType&) = 0;
+
+    /** Queue a job to call invokeOnTimer(). */
+    virtual void
+    queueJob() = 0;
```

Comment:
The only responsibility of `queueJob` is to call `invokeOnTimer`, which itself short-circuits (i.e. cancels) the callback (which is `onTimer` method) if the object is "done". `queueJob` gets a chance to throttle and call `setTimer` instead. Every subtype follows a common pattern for `queueJob`:

```cpp
    if (app_.getJobQueue().getJobCountTotal(jtJOB_TYPE) >
        MAX_JOBS_OF_TYPE
    {
        JLOG(m_journal.debug())
            << "Deferring ClassName timer due to load";
        setTimer();
        return;
    }

    std::weak_ptr<LedgerReplayTask> wptr = shared_from_this();
	app_.getJobQueue().addJob(jtJOB_TYPE, "ClassName", [wptr](Job&) {
        if (auto sptr = wptr.lock(); sptr)
            sptr->invokeOnTimer();
    });
```

Perhaps they can share a common implementation, parameterized by:

- The class name.
- The job category.
- The threshold max number of jobs of that category. One subtype has no limit (`TransactionAcquire`), which is the same as a maximum representable value limit (`std::numeric_limits<std::invoke_result_t<decltype(&JobQueue::getJobCountTotal)(JobQueue)>>::max()`.

Short of a common implementation, we should document that this pattern of implementation is expected.

Reply:
Added a common implementation, to pull #3